pmlc/dialect/pxa/transforms/vectorize.cc

```diff
@@ -0,0 +1,262 @@
+    result.setType(vecType);
+  }
+
+  void vectorizeLoadOp(AffineLoadOp op) {
+    Value operand = op.getMemRef();
+    auto eltType = op.getMemRefType().getElementType();
+    auto vecType = VectorType::get(vectorSize, eltType);
+    OpBuilder builder(op);
+    auto vecOp = builder.create<AffineVectorLoadOp>(
+        op.getLoc(), ArrayRef<Type>{vecType}, operand, op.indices());
+    // TODO: Add support for direct construction with map to VectorLoadOp
+    auto mapAttr = AffineMapAttr::get(op.getAffineMap());
+    vecOp.setAttr(AffineVectorLoadOp::getMapAttrName(), mapAttr);
+    op.replaceAllUsesWith(vecOp.getResult());
+    op.erase();
+  }
+
+  void vectorizeReduceOp(AffineReduceOp op) {
+    Value mem = op.mem();
+    Value vector = op.val();
+    assert(vector.getType().isa<VectorType>());
+    OpBuilder builder(op);
+    auto vecOp = builder.create<AffineVectorReduceOp>(
+        op.getLoc(), ArrayRef<Type>{mem.getType()}, op.agg(), vector, mem,
+        op.map(), op.idxs());
+    op.replaceAllUsesWith(vecOp.getResult());
+    op.erase();
+  }
+
+  void vectorizeOperation(Operation *op) {
+    if (!vectorizedOps.count(op)) {
+      return;
+    }
+    if (auto loadOp = dyn_cast<AffineLoadOp>(op)) {
+      vectorizeLoadOp(loadOp);
+    } else if (auto reduceOp = dyn_cast<AffineReduceOp>(op)) {
+      vectorizeReduceOp(reduceOp);
+    } else {
+      vectorizeScalarOp(op);
+    }
+  }
+
+  LogicalResult vectorize() {
+    mlir::Block *body = loop.getBody();
+    IVLOG(3, "Vectorize: Attempting Vectorizing for BlockArgument: "
+                 << index.getArgNumber());
+
+    auto ranges = loop.getConstantRanges();
+    if (!ranges) {
+      IVLOG(3, " Vectorize: Failed, Requires constant ranges");
+      return failure();
+    }
+
+    auto argNum = index.getArgNumber();
+    if ((*ranges)[argNum] % vectorSize != 0) {
+      IVLOG(3, "Vectorize: Failed, the dimension being vectorized not multiple "
+               "of the number of elements in a register");
+      return failure();
+    }
+
+    auto step = loop.steps().getValue()[argNum].cast<IntegerAttr>().getInt();
+    if (step != 1) {
+      IVLOG(3,
+            "Vectorize: Failed, the steps for the dimension being vectorized "
+            "is not 1");
+      return failure();
+    }
+
+    bool vectorizable = true;
+    body->walk(
+        [&](Operation *op) { vectorizable &= tryVectorizeOperation(op); });
+    if (!vectorizable) {
+      return failure();
+    }
+    if (vectorizedOps.empty()) {
+      // TODO: should we actually fail in this case?  Currently we need to since
+      // we have no cost model
+      IVLOG(3, "Vectorize: Failed, no point in vectorization");
+      return failure();
+    }
+
+    // Preflight complete, do the transform
+    for (auto &op : llvm::make_early_inc_range(body->getOperations())) {
+      vectorizeOperation(&op);
+    }
+    llvm::SmallVector<int64_t, 6> steps;
+    for (auto ia : loop.steps().cast<ArrayAttr>().getValue()) {
```

Comment:
I don't know what `ia` means in this context, can you make a slightly longer var name?